Sources/WendyAgent/Services/ContainerMonitor.swift

```diff
@@ -0,0 +1,264 @@
+                restartPolicy: restartPolicy
+            )
+            logger.debug("Created new container state for \(appName)")
+        } else {
+            containerStates[appName]?.restartPolicy = restartPolicy
+        }
+
+        guard var state = containerStates[appName] else { return }
+
+        // Check if container should be restarted
+        let shouldRestart = shouldRestartContainer(
+            isRunning: isRunning,
+            state: state,
+            exitCode: Int(task?.exitStatus ?? 0)
+        )
+
+        if shouldRestart {
+            logger.debug(
+                "Restarting container",
+                metadata: [
+                    "container": "\(appName)",
+                    "failure-count": "\(state.failureCount)"
+                ]
+            )
+
+            do {
+                // Utilize nerdctl to restart the container
+                _ = try await runNerdctl(["start", appName])
+
+                state.lastRestartTime = Date()
+                state.failureCount += 1
+                containerStates[appName] = state
+
+                logger.debug(
+                    "Container restarted successfully",
+                    metadata: ["container": "\(appName)"]
+                )
+            } catch {
+                logger.error(
+                    "Failed to restart container",
+                    metadata: [
+                        "container": "\(appName)",
+                        "error": "\(error)"
+                    ]
+                )
+            }
+        }
+    }
+
+    /// Determine if a container should be restarted based on its state and policy
+    private func shouldRestartContainer(
+        isRunning: Bool,
+        state: ContainerState,
+        exitCode: Int
+    ) -> Bool {
+        // Don't restart if explicitly stopped by user
+        if state.explicitlyStopped {
+            return false
+        }
+
+        // Container is running fine
+        if isRunning {
+            return false
+        }
+
+        // Container is stopped, check restart policy
+        switch state.restartPolicy {
+        case .no:
+            return false
+
+        case .always, .unlessStopped:
+            // Always restart unless explicitly stopped
+            return true
+
+        case .onFailure(let maxRetries):
+            // Only restart on non-zero exit code, up to max retries
+            if exitCode != 0 && state.failureCount < maxRetries {
+                return true
+            }
+            return false
+        }
+    }
+
+    /// Run nerdctl command for container operations
+    private func runNerdctl(_ args: [String]) async throws -> String {
+        let fullArgs = ["--namespace", "default"] + args
+        logger.info("Running nerdctl command", metadata: ["args": "\(fullArgs.joined(separator: " "))"])
+
+        do {
+            let result = try await Subprocess.run(
+                .name("/usr/local/bin/nerdctl"),
```

Comment:
```suggestion
                .name("nerdctl"),
```

This should also work